### PR TITLE
Dependency Checking

### LINKED ISSUE #73: [FEATURE REQUEST] Dependency checking on scan start
Not immediately clear when dependencies are missing from a scan

"requires" function should check if dependency is installed.
threaded nmap should also "which nmap"

### PR DESCRIPTION
**Pull request by GreaterGoodest**
Adds checks to determine if dependencies are installed before scanning. Notifies user via exception if dependencies are not installed. The purpose of this WIP PR is to receive feedback on re-design of the meets_requirements() function.

Testing needs some work

Closes #73 

## Branching checklist
- [x] There is an issue associated with your PR (bug, feature, etc.. if not, create one)
- [x] Your PR description references the associated issue (i.e. fixes #123)
- [x] Code is in its own branch
- [x] Branch name is related to the PR contents
- [x] PR targets master

## Static analysis checks
- [x] All python files are formatted using black
- [x] All flake8 checks pass
- [x] All existing tests pass

## Documentation
- [x] Documentation about your PR is included in docs/ and the README, as appropriate

## Additional Tests
- [x] New code is unit tested
- [x] New tests pass

### CODE REVIEW

File: pipeline/recon-pipeline.py [hunk 1]
```diff
@@ -298,10 +298,12 @@ def do_scan(self, args):
...
         # each classname corresponds to a potential recon-pipeline command, i.e. AmassScan, GobusterScan ...
         scans = get_scans()
-
         # command is a list that will end up looking something like what's below
         # luigi --module pipeline.recon.web.webanalyze WebanalyzeScan --target abc.com --top-ports 100 --interface eth0
-        command = ["luigi", "--module", scans.get(args.scantype)[0]]
+        try:
+            command = ["luigi", "--module", scans.get(args.scantype)[0]]
+        except TypeError:
+            raise TypeError(style(f"[!] {args.scantype} not installed", fg="bright_red"))
```
Comment (epi052) [inline pipeline/recon-pipeline.py:306]: Why a re-raise instead of a print and return?

File: pipeline/recon/helpers.py [hunk 2]
```diff
@@ -19,6 +21,15 @@ def get_tool_state() -> typing.Union[dict, None]:
         return pickle.loads(tools.read_bytes())
 
 
+def meets_requirements(requirements):
```
Comment (epi052) [inline pipeline/recon/helpers.py:24]: while i like the idea of staying DRY with a helper function, this solution breaks the tab-completion checking of installed commands (not verified, but pretty darn confident).

File: pipeline/recon/helpers.py [hunk 3]
```diff
@@ -19,6 +21,15 @@ def get_tool_state() -> typing.Union[dict, None]:
...
+    """ Determine if tools required to perform task are installed. """
+    tools = get_tool_state()
+
+    for tool in requirements:
+        if not tools.get(tool).get("installed"):
+            raise (Exception(style(f"[!] {tool} is not installed", fg="bright_red")))
```
Comment (epi052) [inline pipeline/recon/helpers.py:30]: - parens surrounding Exception aren't necessary
- RuntimeError is probably a better option for what to raise

consider us calling the function

```
try:
    meets_requirements(['stuff'])
except Exception:
    # handle stuff
```

Literally everything that can possibly go wrong is caught under that, with potentially non-specific exception handling.  Better practice is to not catch the base Exception and be more specific.

File: pipeline/recon/helpers.py (hunk 2 above)
Comment (GreaterGoodest) [reply pipeline/recon/helpers.py:24]: Tab completion appears to still be functional for all contexts

**Comment by GreaterGoodest** [comment]
Need to dig in and figure out why I'm wrecking the pipeline

**Comment by GreaterGoodest** [comment]
Adjusted meets_requirements implementation and it now works properly, need to fix tests

**Comment by GreaterGoodest** [comment]
Tests pass locally, trying to figure out what's going on with pipeline

**Comment by GreaterGoodest** [comment]
Ready for review!

File: pipeline/recon/web/gobuster.py [hunk 4]
```diff
@@ -59,21 +59,14 @@ class GobusterScan(luigi.Task):
     threads = luigi.Parameter(default=defaults.get("threads"))
     wordlist = luigi.Parameter(default=defaults.get("gobuster-wordlist"))
     extensions = luigi.Parameter(default=defaults.get("gobuster-extensions"))
+    requirements = ["recursive-gobuster", "gobuster"]
```
Comment (epi052) [inline pipeline/recon/web/gobuster.py:62]: unless it's handled somewhere else, go is a requirement as well

Comment (GreaterGoodest) [reply pipeline/recon/web/gobuster.py]: This was an oversight, corrected but pipeline fails due to Go version mismatch

File: pipeline/recon/web/aquatone.py [hunk 5]
```diff
@@ -58,21 +58,14 @@ class AquatoneScan(luigi.Task):
 
     threads = luigi.Parameter(default=defaults.get("threads", ""))
     scan_timeout = luigi.Parameter(default=defaults.get("aquatone-scan-timeout", ""))
+    requirements = ["aquatone"]
```
Comment (epi052) [inline pipeline/recon/web/aquatone.py:61]: masscan

File: pipeline/recon/web/gobuster.py [hunk 6]
```diff
@@ -59,21 +59,14 @@ class GobusterScan(luigi.Task):
     threads = luigi.Parameter(default=defaults.get("threads"))
     wordlist = luigi.Parameter(default=defaults.get("gobuster-wordlist"))
     extensions = luigi.Parameter(default=defaults.get("gobuster-extensions"))
+    requirements = ["recursive-gobuster", "go", "gobuster"]
```
Comment (epi052) [inline pipeline/recon/web/gobuster.py:62]: masscan

File: pipeline/recon/web/subdomain_takeover.py [hunk 7]
```diff
@@ -47,21 +47,15 @@ class TKOSubsScan(luigi.Task):
         results_dir: specifes the directory on disk to which all Task results are written *Required by upstream Task*
     """
 
+    requirements = ["go", "tko-subs"]
```
Comment (epi052) [inline pipeline/recon/web/subdomain_takeover.py:50]: masscan

File: pipeline/recon/web/subdomain_takeover.py [hunk 8]
```diff
@@ -177,22 +172,15 @@ class SubjackScan(luigi.Task):
     """
 
     threads = luigi.Parameter(default=defaults.get("threads"))
+    requirements = ["go", "subjack"]
```
Comment (epi052) [inline pipeline/recon/web/subdomain_takeover.py:175]: masscan

File: pipeline/recon/web/waybackurls.py [hunk 9]
```diff
@@ -44,20 +44,14 @@ class WaybackurlsScan(luigi.Task):
         results_dir: specifes the directory on disk to which all Task results are written *Required by upstream Task*
     """
 
+    requirements = ["go", "waybackurls"]
```
Comment (epi052) [inline pipeline/recon/web/waybackurls.py:47]: masscan

File: pipeline/recon/web/webanalyze.py [hunk 10]
```diff
@@ -54,21 +54,14 @@ class WebanalyzeScan(luigi.Task):
     """
 
     threads = luigi.Parameter(default=defaults.get("threads"))
+    requirements = ["go", "webanalyze"]
```
Comment (epi052) [inline pipeline/recon/web/webanalyze.py:57]: masscan

File: pipeline/recon/wrappers.py [hunk 11]
```diff
@@ -111,17 +106,12 @@ class HTBScan(luigi.WrapperTask):
...
 
-    @staticmethod
-    def meets_requirements():
-        """ Reports whether or not this scan's needed tool(s) are installed or not """
-        needs = ["aquatone", "masscan", "recursive-gobuster", "searchsploit", "gobuster", "webanalyze"]
-        tools = get_tool_state()
-
-        if tools:
-            return all([tools.get(x).get("installed") is True for x in needs])
+    requirements = ["aquatone", "masscan", "recursive-gobuster", "searchsploit", "gobuster", "webanalyze"]
```
Comment (epi052) [inline pipeline/recon/wrappers.py:109]: go

File: pipeline/recon/wrappers.py [hunk 12]
```diff
@@ -28,28 +28,23 @@ class FullScan(luigi.WrapperTask):
...
-            "subjack",
-            "gobuster",
-            "webanalyze",
-            "waybackurls",
-        ]
-        tools = get_tool_state()
-
-        if tools:
-            return all([tools.get(x).get("installed") is True for x in needs])
+    requirements = [
```
Comment (epi052) [inline pipeline/recon/wrappers.py:31]: go

**Pull request merged**